### ISSUE TITLE
signal.decimate() introduces nasty phase-shift

### ISSUE THREAD

**Issue by ankostis** [scipy/scipy#4929]
The `decimate()` function by default adds a strong phase shift to the data, apparently because it used a single pass (low pass filter and not ``scipy.signal.filtfilt()`` 

The problem has been reported earlier[1][2], and the suggested workaround is to clone-and-adapt the code of the function.

I tried this on my sample data and got satisfactory results:
![decimatephaseshift](https://cloud.githubusercontent.com/assets/501585/7927242/c1a5ad14-08dd-11e5-9f60-807d08175eec.png)

May I provide a PR to modify `decimate()` to accept a new boolean *kw-arg* `filtfilt` to do the above?
And possibly another one, `filtfilt_kws` ??

[1] http://mail.scipy.org/pipermail/scipy-user/2011-April/029063.html
[2] http://scipy-user.10969.n7.nabble.com/SciPy-User-Decimate-and-filtfilt-td20183.html

**Comment by e-q** [comment]
+1 from me

While ``decimate `` is up for discussion, my two cents:

The current IIR filter chosen is a type-1 Chebyshev, but with only 0.05dB allowed passband ripple. At that point, I feel it may be more advantageous use a type-2 Cheby filter, which naturally has a flat passband, and has a notch in the stopband which can be aligned with the nyquist frequency. 

Alternatively, if we want to make the filtering more user-controllable, this seems like a natural place to use something the  ``[filt]ord`` functions, (like ``cheb2ord``, ``ellipord``, etc.) to let the user specify how steep they want the cutoff / how flat they want the passband / what kind of filter they want to use (along with some sensible defaults).

**Comment by FRidh** [comment]
Thanks for bringing this up, I was unaware decimate introduced this phase shift.

About the keyword. I don't think there are (or will be!) any other options than `lfilter` or `filtfilt` and therefore I think a better name would be  `zero_phase=True` for `filtfilt`.

`filtfilt` has some keywords, and you might want to pass those as well. One solution would indeed be `filtfilt_kws`.

Another would be using something like `mode='lfilter'` or `mode=filtfilt` or `mode=some_func`. Why? While I don't expect one would use any other function than `lfilter` or `filtfilt` this method would allow you so and this way you could also add keywords through the use of `functools.partial`.

**Comment by ankostis** [comment]
I agree, the power of `partial()` is frequently under-appreciated.

But in this case i would prefer the 1st suggestion, the a `zero_phase` +`filtfilt_kws`, because it looks simpler; for more complex situations, decimate is a small and you can easily clone and adapt.

The problem is that a 2-pass filtering needs lower-order filter-coefficients.
In my case i have found the following defaults for the `n` parameter to work:

filter-type | # of passes | filter-order
------------ | ---------------| -------------
`fir` |  `single-pass`| 30
`iir` |  `single-pass`| 8
`fir` |  `2-pass`| 26
`iir` |  `2-pass`| 6

But i would expect more trials on this.

**Comment by rgommers** [comment]
@e-q the ``cheby1`` with 0.05 dB passband seems to have been chosen to match Matlab: http://nl.mathworks.com/help/signal/ref/decimate.html

Making this more configurable without complicating the API with even more keywords looks difficult. Instead of more strings or a callable, the simplest might be to just accept one of the ``lti`` subclasses or a tuple ``(b, a)``.

**Comment by rgommers** [comment]
@ankostis a PR adding ``filtfilt`` support sounds good. ``zero_phase`` makese sense. Not 100% sure about ``filtfilt_kws``, that makes the API a bit complex.

The problem with this function is that it's basically a two-liner:

    b, a = construct_filter(n, q, kwds)
    res = lowpass_filter(b, a, x, other_kwds)

Adding lots of keywords to ``decimate`` to tweak the behavior of ``construct_filter`` or ``lowpass_filter`` quickly makes ``decimate`` less usable than a reimplementation by hand.

**Comment by FRidh** [comment]
Something like `decimate(x, q, filt, axis=-1, zero_phase=False)` where `filt` is an instance of a subclass of `lti`? 
This would be backwards incompatible though.

Unless, the `zero_phase` keyword would be added, and `ftype` could be an instance of a subclass of `lti` besides the current `iir` and `fir` options.

**Comment by rgommers** [comment]
Breaking backwards compat isn't needed I think. Adding ``zero_phase=False`` at the end, and making ``ftype`` accept ``signal.TransferFunction`` is what I meant.

**Comment by FRidh** [comment]
I added the `zero_phase` keyword as well as the ability for `ftype` to accept an instance of `lti`.
Please have a look at https://gist.github.com/ba60084570d05c91e9bd

You see two versions, the one where I use `b, a` and the other where I use an instance of `system` internally. While right now it doesn't matter at all which is used, I think for in the future it's best to already use the system representation. As soon as we then have a `SecondOrderSections` and more functions support `sos` it's rather straight forward to convert.

**Comment by e-q** [comment]
@FRidh, I've used the code you posted for a little bit, and it seems fine to me. Why not open a PR for it?

**Comment by e-q** [comment]
I've taken the liberty of creating a PR based on @FRidh gist, since I want to see this happen.

**Comment by haudren** [comment]
Thank you for this, I just stumbled on this problem today :+1: 

I would like to remark that the behaviour is quite inconsistent when using a non-zero constant signal:
```python
from scipy.signal import decimate
import numpy as np

A = np.array([1.]*200)
print decimate(A, 20)
```
Returns:
```python
[  1.56761563e-11   1.81198225e-03   1.40286212e-01   8.08457647e-01
   1.17621428e+00   9.02814439e-01   1.02701779e+00   1.00592634e+00
   9.69387837e-01   1.02763790e+00]
```
While Octave returns a constant size-10 vector of value 0.98855. Matlab goes a bit further: it checks that the signal length is divisible by the downsampling factor. If it is, it simply returns a slice [:q:]. Else, same as Octave.

**Comment by FRidh** [comment]
@haudren 
`filtfilt` is known to cause edge-effects:
https://mail.scipy.org/pipermail/scipy-user/2011-April/029063.html

We might need to add an issue for that here in the tracker.

**Comment by haudren** [comment]
@FRidh 
There may be an incomprehension here. In the code above, I used the actual implement of signal.decimate (that use lfilt). I tried using the code proposed in the PR, that uses filtfilt and obtained a very reasonable result:
```python
[ 0.98855104,  0.98855227,  0.98855194,  0.98855081,  0.9885514 ,
        0.98855173,  0.98855112,  0.98855203,  0.98855287,  0.98855334]
```

**Comment by WarrenWeckesser** [comment]
+1 for providing an option to use `filtfilt` in `decimate`.

@FRIdh wrote:

> filtfilt is known to cause edge-effects:
https://mail.scipy.org/pipermail/scipy-user/2011-April/029063.html

The issue mentioned there was fixed in 2011, not long after that email.  Bug or not, it is an unavoidable fact that *all* filters introduce edge effects.  There are several heuristic tricks and techniques that attempt to minimize the edge effects, but these are all mathematically questionable.  For example, padding extends the signal with fictitious data.  It might make a plot look better, but a rigorous justification is, at best, handwavy.

@haudren  wrote:

> Matlab goes a bit further: it checks that the signal length is divisible by the downsampling factor. If it is, it simply returns a slice [:q:].

Depending on what you are going to do with the result, that could be a bad idea.  If the input signal has frequency content above the downsampled Nyquist frequency, simply slicing the signal will result in aliasing.  It is essential to use a low-pass filter before slicing to avoid aliasing.

**Comment by e-q** [comment]
This issue affects FIR decimation as well, unsurprisingly. As a comparison, MATLAB uses `filtfilt` for IIR decimation, but a different strategy for FIR. From [the MATLAB docs](http://www.mathworks.com/help/signal/ref/decimate.html):

> When using the FIR filter, decimate filters the input sequence in only one direction. This conserves memory and is useful for working with long sequences. In the IIR case, decimate applies the filter in forward and reverse directions using filtfilt to remove phase distortion. This in effect doubles the filter order. In both cases, the function minimizes transient effects at both ends of the signal by matching endpoint conditions.

It then cites a reference for the algorithms used:
> [1] Digital Signal Processing Committee of the IEEE Acoustics, Speech, and Signal Processing Society, eds. Programs for Digital Signal Processing. New York: IEEE Press, 1979, chap. 8.

I think the memory argument is fair point; I'll see if I can track down this reference and whip something up. This could possibly be included as an extension to #5392

Just as a point of reference, here how scipy v0.16.1's  `decimate` does at reproducing the FIR example from the MATLAB page

![firdecimation](https://cloud.githubusercontent.com/assets/6894112/12767040/b23d466c-c9bb-11e5-81eb-e1f17ab5aa34.png)

This was created with 
```
r = 13
n = np.arange(16, 366)
x = np.sin(2*np.pi*n/153.0) + np.cos(2*np.pi*n/127.0)
y = sig.decimate(x, r, 82, 'fir')

plt.plot(np.arange(len(n)), x, 'bo')
plt.stem(np.arange(len(n))[::r], y, linefmt='r-', markerfmt='ro', basefmt='k-')
plt.legend(['Original', 'Decimated'])
plt.xlabel('Sample Number')
plt.ylabel('Signal')
plt.show()
```
